### DISCUSSION TITLE
Custom Appender documentation possible errors

### DISCUSSION THREAD

**Discussion by keithrbennett** [reidmorrison/semantic_logger#227] [Q&A] [answered]
1) The example code at https://logger.rocketjob.io/custom_appenders.html includes a call to `include_message?`, but I didn't see that method defined anywhere. I have removed it from my custom appender, but am wondering, is this important, and should I implement it myself? If so, what should it do?

2) I found that I needed to provide the keyword name for the code to work:

```ruby
      super(level: level, &block)
```
----

I am writing an appender that I can query in unit tests. So far I have this, in case it's useful to you, or you have any suggestions:

```ruby
  class LogEventAppender < SemanticLogger::Subscriber

    attr_reader :events

    def initialize(level = nil, &block)
      super(level: level, &block)
      @events = []
    end

    # Display the log struct and the text formatted output
    def log(log)
      @events << log if level_index > (log.level_index || 0)
    end

    def clear
      @events.clear
    end
  end
```

Then, during rspec initialization, in a `logging_spec_helper.rb`, I do this:

```ruby
  def event_appender
    $semantic_logger_event_appender ||= LogEventAppender.new
  end

  def events
    SemanticLogger.flush
    event_appender.events
  end

  def setup_logging(rspec_config)
    rspec_config.before(:each) do
      SemanticLogger.flush
      event_appender.clear
  end
  ...
end
```

Then, in a given test, I can call `LoggingSpecHelper.events` and get only the log events logged during that test.

**Comment by reidmorrison** [comment]
The appender you are looking for already exists, see the section `Writing Tests` on this page for information on how to write tests to verify logging events: https://logger.rocketjob.io/api.html.

**Reply by keithrbennett** [reply]
Regarding the CaptureLogEvents class, there are two problems I have with it that have motivated me to write my own test appender:

1) If `events`  is called before any log events have been processed, it returns nil. I would prefer it return an empty array, which of course, could be easily done by initializing it in the constructor.
2) I want to start with an empty log for each unit test. I would prefer not to go to a level as high as creating a new appender and replacing the existing one, but rather to clear the data store of the existing logger. My test appender has a `clear` method that clears the `events` array, and is called in a global `before` hook.

**Comment by reidmorrison** [comment, marked as answer]
Thank you, Documentation updated.

**Comment by reidmorrison** [comment]
The documentation for the new Minitest extension for testing log messages:  https://logger.rocketjob.io/testing.html

@keithrbennett, since it looks like you use rspec, would you be interested in creating the rspec equivalent of the following minitest extensions just added to SemanticLogger on the master branch?

Would be great to include the rspec equivalent when the above minitest helper is published in a gem. It is just in master for now. If you want to try it locally, update Gemfile with:
`gem "semantic_logger", github: "reidmorrison/semantic_logger"`

**Reply by keithrbennett** [reply]
@reidmorrison Unfortunately, I am super busy for the next few weeks. I took a look at the master branch; could you tell me which files contain the changes ("the following minitest extensions")? Sorry if it's obvious and I'm missing it.
